lib/falcon/adapters/rack.rb

```diff
@@ -105,6 +105,7 @@ def unwrap_request(request, env)
 				# https://tools.ietf.org/html/rfc7239#section-5.4
 				# https://github.com/rack/rack/issues/1310
 				env[HTTP_X_FORWARDED_PROTO] ||= request.scheme
+				env[RACK_URL_SCHEME] = env[HTTP_X_FORWARDED_PROTO]
```

Comment:
Did we set `RACK_URL_SCHEME` anywhere else or did I just completely miss it?

Also, is this checked by rack linter?

Reply:
`RACK_URL_SCHEME` defaults to `request.scheme`, however that's not accurate in cases where we're listening on HTTP but have SSL terminated upstream.

Reply:
If the user sent the `X-Forwarded-Proto: https` could that break the application that's running on `http`? Just wondering if there are security implications to this.

Reply:
For context, here's how puma sets `RACK_URL_SCHEME`:

```ruby
env[RACK_URL_SCHEME] = default_server_port(env) == PORT_443 ? HTTPS : HTTP
```

Here's `default_server_port`:

```ruby
def default_server_port(env)
  if ['on', HTTPS].include?(env[HTTPS_KEY]) || env[HTTP_X_FORWARDED_PROTO].to_s[0...5] == HTTPS || env[HTTP_X_FORWARDED_SCHEME] == HTTPS || env[HTTP_X_FORWARDED_SSL] == "on"
    PORT_443
  else
    PORT_80
  end
end
```

In theory if the application was exposed directly, you could bypass middleware to force ssl by sending `X-Forwarded-Proto: https`. Whether or not that has security implications I don't know, since it could only affect the current user's session. 

I'm not certain on this, but I would think it's up to nginx, haproxy, or whatever is used in front of the application to set or reject these headers. Here's how Heroku handles it:

* https://devcenter.heroku.com/articles/http-routing#heroku-headers

Reply:
@ioquatix the purpose of the scheme check is to protect the user from sending information over unauthenticated connections. If the user provides the `X-Forwarded-Proto: https` it will only impact them. I don't think this is a security concern.